ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/operations/AttestationData.java

```diff
@@ -141,6 +144,24 @@ public UnsignedLong getEarliestSlotForProcessing() {
     return max(getSlot().plus(UnsignedLong.ONE), getTarget().getEpochStartSlot());
   }
 
+  public boolean canIncludeInBlockAtSlot(final UnsignedLong slot) {
+    return notPriorToPreviousEpoch(slot)
+        && notSubmittedTooFarInHistory(slot)
+        && notSubmittedTooQuickly(slot);
+  }
+
+  public boolean notPriorToPreviousEpoch(final UnsignedLong slot) {
+    return compute_epoch_at_slot(getSlot()).plus(ONE).compareTo(compute_epoch_at_slot(slot)) >= 0;
+  }
+
+  public boolean notSubmittedTooFarInHistory(final UnsignedLong slot) {
```

Comment:
It seems to me that this check makes the notPriorToPreviousEpoch check redundant. Am I missing an edge case?

Reply:
The spec has:
```
assert data.target.epoch in (get_previous_epoch(state), get_current_epoch(state))
assert data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= data.slot + SLOTS_PER_EPOCH
```
so this is another issue with missing `MIN_ATTESTATION_INCLUSION_DELAY`.  You have to include it within `SLOTS_PER_EPOCH` of when it's first eligible to include, not from the attestation's slot.

Reply:
hmm, stil not entirely sure I'm interpreting that right - this may need coffee...

Reply:
I'm convinced that the requirement that attestation.slot is within SLOTS_PER_EPOCH of state.slot means that it can't be from before the previous epoch. Combined with the check for the state being at least `MIN_ATTESTATION_INCLUSION_DELAY` after the attestation slot, it also can't be a future block so winds up having to be either the current or previous epoch.  I've removed the redundant check.